<a href="https://colab.research.google.com/github/kunalshenoy1604/CODSOFT/blob/main/Task2_SMS_SpamDetection_by_KunalShenoy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IMPORTING THE NECESSARY LIBRARIES

In [45]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OrdinalEncoder

MOUNTING GOOGLE DRIVE & READING SPAM.CSV FILE

In [46]:
df = pd.read_csv("/content/drive/MyDrive/spam.csv", encoding='ISO-8859-1')
df

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

DROPPING UNNECESSARY COLUMNS CREATED DUE TO ENCODING

In [47]:
df.drop(columns=['Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], axis=1, inplace=True)
df.head(10)

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
5  spam  FreeMsg Hey there darling it's been 3 week's n...
6   ham  Even my brother is not like to speak with me. ...
7   ham  As per your request 'Melle Melle (Oru Minnamin...
8  spam  WINNER!! As a valued network customer you have...
9  spam  Had your mobile 11 months or more? U R entitle...

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   v1      5572 non-null   object
 1   v2      5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [49]:
df.columns

Index(['v1', 'v2'], dtype='object')

In [50]:
df.shape

(5572, 2)

THERE IS NO NULL VALUES IN THE DATASET AND HENCE NO IMPUTATION IS NECESSARY

In [51]:
df.isnull().sum()

v1    0
v2    0
dtype: int64

CONVERTING SPAM & HAM FROM DATASET TO BINARY NUMBERS (1 & 0) RESPECTIVELY

In [52]:
df['v1'] = df['v1'].replace({'spam': 1, 'ham': 0})
df.head(10)

v1                                                 v2
0   0  Go until jurong point, crazy.. Available only ...
1   0                      Ok lar... Joking wif u oni...
2   1  Free entry in 2 a wkly comp to win FA Cup fina...
3   0  U dun say so early hor... U c already then say...
4   0  Nah I don't think he goes to usf, he lives aro...
5   1  FreeMsg Hey there darling it's been 3 week's n...
6   0  Even my brother is not like to speak with me. ...
7   0  As per your request 'Melle Melle (Oru Minnamin...
8   1  WINNER!! As a valued network customer you have...
9   1  Had your mobile 11 months or more? U R entitle...

In [53]:
print(df.v1.value_counts(normalize=True))

0    0.865937
1    0.134063
Name: v1, dtype: float64


CLEAR INDICATION THAT 86.59% SMS Messages are NOT SPAM (HAM)

SELECTING V2 i.e THE SMS MESSAGE AS A FEATURE VARIABLE WHEREAS SELECTING V1 AS A TARGET VARIABLE

In [54]:
X=df['v2'] #Feature Variables
y=df['v1'] #target variable

In [55]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.3, random_state=10)

USING TF-IDF TO VECTORIZE THE TEXT MESSAGES  

In [56]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

USING SUPPORT VECTOR MACHINE CLASSIFIER FOR CLASSIFICATION

In [57]:
from sklearn.svm import SVC
classifier=SVC(kernel='linear',random_state=0)
classifier.fit(X_train_tfidf,y_train)

SVC(kernel='linear', random_state=0)

In [61]:
y1_pred=classifier.predict(X_test_tfidf)
y1_pred

array([0, 0, 0, ..., 0, 0, 0])

In [63]:
from sklearn import metrics
print("Accuracy:-")
print(metrics.accuracy_score(y_test,y1_pred))

Accuracy:-
0.9886363636363636


ACCURACY OF THE CLASSIFICATION MODEL USING SUPPORT VECTOR MACHINE = 98.86%

USING NAIVE BAYES CLASSIFIER FOR CLASSIFICATION

In [65]:
from sklearn.naive_bayes import MultinomialNB
classifier=MultinomialNB()
classifier.fit(X_train_tfidf,y_train)

MultinomialNB()

In [69]:
y2_pred = classifier.predict(X_test_tfidf)
y2_pred

array([0, 0, 0, ..., 0, 0, 0])

In [70]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [72]:
accuracy = accuracy_score(y_test, y2_pred)
accuracy

0.9563397129186603

ACCURACY OF THE CLASSIFICATION MODEL USING NAIVES BAYES CLASSIFIER = 95.63%

In [74]:
confusion = confusion_matrix(y_test, y2_pred)
confusion

array([[1446,    0],
       [  73,  153]])

# CONCLUSION:
# ACCURACY (SVM) : 98.86%
#           
# ACCURACY (NAIVE BAYES) : 95.63%